bionic/protocols.py

```diff
@@ -48,33 +48,69 @@ def check_is_like_protocol(obj):
 
 class BaseProtocol:
     def validate(self, value):
+        """
+        Checks if a value is valid for this protocol. Throws an exeception if the value
+        is invalid; otherwise does nothing.
+        """
         pass
 
     def value_is_valid(self, value):
+        """Indicates whether the passed value is valid for this protocol."""
         try:
             self.validate(value)
         except Exception:
             return False
         return True
 
     def get_fixed_file_extension(self):
+        """
+        Returns a file extension identifying this protocol. This value will be appended
+        to the name of any file written by the protocol, and may be used to determine
+        whether a file can be read by the protocol.
+
+        This string should be unique, not shared with any other protocol. By
+        convention, it doesn't include an initial period, but may include periods in
+        the middle.  (For example, `"csv"`, and `"csv.zip"` would both be sensible
+        file extensions.)
+
+        If a protocol uses different file extensions for different input values, it
+        should implement ``file_extension_for_value`` instead.
+        """
         raise NotImplementedError()
 
     def file_extension_for_value(self, value):
+        """
+        Like ``get_fixed_file_extension``, but may provide different extensions
+        depending on the value to be serialized.
+        """
         return self.get_fixed_file_extension()
 
     def can_read_file_extension(self, extension):
+        """Indicates whether a file extension can be read by this protocol."""
         return extension == self.get_fixed_file_extension()
 
     def supports_value(self, value):
+        """Indicates whether a value can be serialized by this protocol."""
         return any(isinstance(value, type_) for type_ in self.get_all_supported_types())
 
     def write(self, value, path):
+        """Serializes the object ``value`` to the pathlib path ``path``."""
         raise NotImplementedError()
 
     def read(self, path):
+        """Deserializes an object from the pathlib path ``path``, and returns it."""
         raise NotImplementedError()
 
+    def read_with_extension(self, path, extension):
+        """
+        Deserializes an object from the pathlib path ``path``, and returns it.
+
+        May use the file extension ``extension`` to determine what deserialization
```

Comment:
nit: is file extension ``extension`` a typo?

Reply:
No, I was trying to use the same pattern as "pathlib path ``path``" above. If I just wrote "the file ``extension``" , that would sound to me like ``extension`` was a file.